File: src/moonlight/browser.py
"""Playwright wrapper — async context-managed Chromium session for scenarios."""
from __future__ import annotations

from pathlib import Path
from urllib.parse import urljoin

from playwright.async_api import Browser as PWBrowser, Page, async_playwright


class Browser:
    def __init__(
        self,
        base_url: str,
        viewport: dict[str, int] | None = None,
        headless: bool = True,
        default_timeout_ms: int = 15000,
        storage_state: Path | str | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/") + "/"
        self.viewport = viewport or {"width": 1280, "height": 720}
        self.headless = headless
        self.default_timeout_ms = default_timeout_ms
        self.storage_state = Path(storage_state) if storage_state else None
        self._pw = None
        self._browser: PWBrowser | None = None
        self._context = None
        self.page: Page | None = None

    async def __aenter__(self) -> "Browser":
        self._pw = await async_playwright().start()
        self._browser = await self._pw.chromium.launch(headless=self.headless)
        ctx_kwargs = {"viewport": self.viewport}
        if self.storage_state and self.storage_state.exists():
            ctx_kwargs["storage_state"] = str(self.storage_state)
        self._context = await self._browser.new_context(**ctx_kwargs)
        self._context.set_default_timeout(self.default_timeout_ms)
        self.page = await self._context.new_page()

        # Per-step rolling buffers — runner snapshots + clears these around each step.
        self.console_messages: list[dict] = []
        self.failed_requests: list[dict] = []
        self.page_errors: list[str] = []

        self.page.on("console", lambda msg: self.console_messages.append({
            "type": msg.type,
            "text": msg.text[:500],
            "location": f"{msg.location.get('url', '')}:{msg.location.get('lineNumber', '')}",
        }))
        self.page.on("pageerror", lambda err: self.page_errors.append(str(err)[:500]))
        self.page.on("requestfailed", lambda req: self.failed_requests.append({
            "url": req.url[:300],
            "method": req.method,
            "failure": (req.failure or "")[:200],
        }))
        self.page.on("response", self._on_response)
        return self

    def _on_response(self, response):
        # Record 4xx/5xx so the report can flag failed API calls.
        try:
            if response.status >= 400:
                self.failed_requests.append({
                    "url": response.url[:300],
                    "method": response.request.method,
                    "status": response.status,
                })
        except Exception:
            pass

    def snapshot_events(self) -> dict:
        """Return current event buffers and clear them for the next step."""
        snap = {
            "console": list(self.console_messages),
            "failed_requests": list(self.failed_requests),
            "page_errors": list(self.page_errors),
        }
        self.console_messages.clear()
        self.failed_requests.clear()
        self.page_errors.clear()
        return snap

    async def __aexit__(self, exc_type, exc, tb) -> None:
        if self._browser:
            await self._browser.close()
        if self._pw:
            await self._pw.stop()

    def _resolve(self, url: str) -> str:
        if url.startswith(("http://", "https://")):
            return url
        return urljoin(self.base_url, url.lstrip("/"))

    async def goto(self, url: str) -> None:
        try:
            await self.page.goto(self._resolve(url), wait_until="networkidle", timeout=self.default_timeout_ms)
        except Exception:
            # Fall back to load — some pages (Google) keep open connections forever.
            await self.page.goto(self._resolve(url), wait_until="load", timeout=self.default_timeout_ms)

    def _locator(self, step: dict):
        """Resolve a YAML step into a Playwright Locator.

        Step may carry exactly one of: selector, role (+ optional name),
        label, text, placeholder, test_id, alt_text, title.
        """
        exact = bool(step.get("exact", False))
        if step.get("selector"):
            return self.page.locator(step["selector"])
        if step.get("role"):
            kwargs: dict = {}
            if step.get("name") is not None:
                kwargs["name"] = step["name"]
            if exact:
                kwargs["exact"] = True
            return self.page.get_by_role(step["role"], **kwargs)
        if step.get("label") is not None:
            return self.page.get_by_label(step["label"], exact=exact)
        if step.get("text") is not None:
            return self.page.get_by_text(step["text"], exact=exact)
        if step.get("placeholder") is not None:
            return self.page.get_by_placeholder(step["placeholder"], exact=exact)
        if step.get("test_id") is not None:
            return self.page.get_by_test_id(step["test_id"])
        if step.get("alt_text") is not None:
            return self.page.get_by_alt_text(step["alt_text"], exact=exact)
        if step.get("title") is not None:
            return self.page.get_by_title(step["title"], exact=exact)
        raise ValueError(
            f"step has no locator (need one of: selector, role, label, text, placeholder, test_id, alt_text, title): {step!r}"
        )

    async def do_click(self, step: dict) -> None:
        await self._locator(step).click()

    async def do_fill(self, step: dict) -> None:
        await self._locator(step).fill(step.get("value", ""))

    async def do_select(self, step: dict) -> None:
        await self._locator(step).select_option(step.get("value"))

    async def do_press(self, step: dict) -> None:
        await self._locator(step).press(step["key"])

    async def do_wait_for(self, step: dict) -> None:
        await self._locator(step).wait_for()

    async def wait_ms(self, ms: int) -> None:
        await self.page.wait_for_timeout(ms)

    async def scroll(self, y: int) -> None:
        await self.page.evaluate(f"window.scrollBy(0, {int(y)})")

    async def screenshot(self, path: Path, full_page: bool = False) -> Path:
        path.parent.mkdir(parents=True, exist_ok=True)
        await self.page.screenshot(path=str(path), full_page=full_page)
        return path

    async def save_storage_state(self, path: Path) -> Path:
        path.parent.mkdir(parents=True, exist_ok=True)
        await self._context.storage_state(path=str(path))
        return path

File: internal/browser/browser.go
// Package browser wraps playwright-go into a tiny ergonomic API matched to
// our YAML step actions: goto/click/fill/select/press/wait_for/wait_ms/scroll/screenshot.
package browser

import (
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/playwright-community/playwright-go"

	"github.com/krisk248/moonlight/internal/scenario"
)

// (imports above are intentionally kept compact; filepath is now used by Download)


type Browser struct {
	pw       *playwright.Playwright
	browser  playwright.Browser
	context  playwright.BrowserContext
	page     playwright.Page
	baseURL  string

	// Per-step event buffers — runner snapshots these between steps.
	ConsoleMessages []ConsoleEntry
	FailedRequests  []NetworkEntry
	PageErrors      []string
}

type ConsoleEntry struct {
	Type     string `json:"type"`
	Text     string `json:"text"`
	Location string `json:"location"`
}

type NetworkEntry struct {
	URL     string `json:"url"`
	Method  string `json:"method"`
	Status  int    `json:"status,omitempty"`
	Failure string `json:"failure,omitempty"`
}

type Opts struct {
	BaseURL          string
	Viewport         scenario.Viewport
	Headless         bool
	DefaultTimeoutMS int
	StorageStatePath string
}

func New(opts Opts) (*Browser, error) {
	if opts.Viewport.Width == 0 {
		opts.Viewport = scenario.Viewport{Width: 1280, Height: 720}
	}
	if opts.DefaultTimeoutMS == 0 {
		opts.DefaultTimeoutMS = 15000
	}

	pw, err := playwright.Run()
	if err != nil {
		return nil, fmt.Errorf("playwright start: %w", err)
	}
	br, err := pw.Chromium.Launch(playwright.BrowserTypeLaunchOptions{
		Headless: playwright.Bool(opts.Headless),
	})
	if err != nil {
		pw.Stop()
		return nil, fmt.Errorf("launch chromium: %w", err)
	}

	ctxOpts := playwright.BrowserNewContextOptions{
		Viewport: &playwright.Size{Width: opts.Viewport.Width, Height: opts.Viewport.Height},
	}
	if opts.StorageStatePath != "" {
		if _, err := os.Stat(opts.StorageStatePath); err == nil {
			ctxOpts.StorageStatePath = playwright.String(opts.StorageStatePath)
		}
	}
	ctx, err := br.NewContext(ctxOpts)
	if err != nil {
		br.Close()
		pw.Stop()
		return nil, fmt.Errorf("new context: %w", err)
	}
	ctx.SetDefaultTimeout(float64(opts.DefaultTimeoutMS))

	page, err := ctx.NewPage()
	if err != nil {
		ctx.Close()
		br.Close()
		pw.Stop()
		return nil, fmt.Errorf("new page: %w", err)
	}

	b := &Browser{
		pw:      pw,
		browser: br,
		context: ctx,
		page:    page,
		baseURL: strings.TrimRight(opts.BaseURL, "/") + "/",
	}

	page.OnConsole(func(msg playwright.ConsoleMessage) {
		loc := msg.Location()
		b.ConsoleMessages = append(b.ConsoleMessages, ConsoleEntry{
			Type:     msg.Type(),
			Text:     truncate(msg.Text(), 500),
			Location: fmt.Sprintf("%s:%d", loc.URL, loc.LineNumber),
		})
	})
	page.OnPageError(func(err error) {
		b.PageErrors = append(b.PageErrors, truncate(err.Error(), 500))
	})
	page.OnRequestFailed(func(req playwright.Request) {
		failure := ""
		if f := req.Failure(); f != nil {
			failure = f.Error()
		}
		b.FailedRequests = append(b.FailedRequests, NetworkEntry{
			URL:     truncate(req.URL(), 300),
			Method:  req.Method(),
			Failure: truncate(failure, 200),
		})
	})
	page.OnResponse(func(resp playwright.Response) {
		if resp.Status() >= 400 {
			b.FailedRequests = append(b.FailedRequests, NetworkEntry{
				URL:    truncate(resp.URL(), 300),
				Method: resp.Request().Method(),
				Status: resp.Status(),
			})
		}
	})

	return b, nil
}

// Page exposes the underlying Playwright page for callers (like the dom
// package) that need raw access.
func (b *Browser) Page() playwright.Page { return b.page }

func (b *Browser) Close() {
	if b.context != nil {
		b.context.Close()
	}
	if b.browser != nil {
		b.browser.Close()
	}
	if b.pw != nil {
		b.pw.Stop()
	}
}

func (b *Browser) resolve(u string) string {
	if strings.HasPrefix(u, "http://") || strings.HasPrefix(u, "https://") {
		return u
	}
	base, _ := url.Parse(b.baseURL)
	rel, _ := url.Parse(strings.TrimLeft(u, "/"))
	return base.ResolveReference(rel).String()
}

// locator resolves a Step into a Playwright Locator.
func (b *Browser) locator(s scenario.Step) (playwright.Locator, error) {
	switch {
	case s.Selector != "":
		return b.page.Locator(s.Selector), nil
	case s.Role != "":
		opts := playwright.PageGetByRoleOptions{}
		if s.RoleName != "" {
			opts.Name = s.RoleName
		}
		return b.page.GetByRole(*playwrightAriaRole(s.Role), opts), nil
	case s.Label != "":
		return b.page.GetByLabel(s.Label), nil
	case s.Text != "":
		return b.page.GetByText(s.Text), nil
	}
	return nil, fmt.Errorf("step has no locator: %+v", s)
}

func playwrightAriaRole(s string) *playwright.AriaRole {
	r := playwright.AriaRole(s)
	return &r
}

// ----- step dispatchers (called by the runner) ------------------------------

func (b *Browser) Goto(u string) error {
	_, err := b.page.Goto(b.resolve(u), playwright.PageGotoOptions{
		WaitUntil: playwright.WaitUntilStateNetworkidle,
		Timeout:   playwright.Float(15000),
	})
	if err != nil {
		// Fall back to load — some apps keep XHR open forever.
		_, err = b.page.Goto(b.resolve(u), playwright.PageGotoOptions{
			WaitUntil: playwright.WaitUntilStateLoad,
		})
	}
	return err
}

func (b *Browser) Click(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	return loc.Click()
}

func (b *Browser) Fill(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	return loc.Fill(s.Value)
}

func (b *Browser) Press(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	return loc.Press(s.Key)
}

func (b *Browser) SelectOption(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	_, err = loc.SelectOption(playwright.SelectOptionValues{Values: &[]string{s.Value}})
	return err
}

func (b *Browser) WaitFor(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	return loc.WaitFor()
}

func (b *Browser) WaitMS(ms int) {
	time.Sleep(time.Duration(ms) * time.Millisecond)
}

// WaitForNetworkidle blocks until the page has had no network activity for
// 500ms — the standard "page is done loading" heuristic.
func (b *Browser) WaitForNetworkidle() error {
	return b.page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
		State: playwright.LoadStateNetworkidle,
	})
}

// SetStepTimeout temporarily overrides the per-step Playwright timeout for
// the upcoming action. Restored to the context's default at the next call.
func (b *Browser) SetStepTimeout(ms int) {
	if ms <= 0 {
		return
	}
	b.context.SetDefaultTimeout(float64(ms))
}

// ResetTimeout restores the global default timeout the runner gave us.
func (b *Browser) ResetTimeout(globalMS int) {
	b.context.SetDefaultTimeout(float64(globalMS))
}

// Upload sets an <input type=file> to a local path on disk.
func (b *Browser) Upload(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	path := s.Path
	if path == "" {
		return fmt.Errorf("upload_file step missing 'path'")
	}
	return loc.SetInputFiles(path)
}

// Download registers a download listener, clicks the trigger, waits for the
// download to complete, then saves it to disk at the given path.
func (b *Browser) Download(s scenario.Step) error {
	loc, err := b.locator(s)
	if err != nil {
		return err
	}
	saveTo := s.SaveTo
	if saveTo == "" {
		saveTo = s.Path
	}
	if saveTo == "" {
		return fmt.Errorf("download_file step missing 'save_to' (or 'path')")
	}

	// page.ExpectDownload runs the trigger inside a closure that arms the listener.
	dl, err := b.page.ExpectDownload(func() error {
		return loc.Click()
	})
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(saveTo), 0o755); err != nil {
		return err
	}
	return dl.SaveAs(saveTo)
}

func (b *Browser) Scroll(y int) error {
	_, err := b.page.Evaluate(fmt.Sprintf("window.scrollBy(0, %d)", y))
	return err
}

func (b *Browser) Screenshot(path string, fullPage bool) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	_, err := b.page.Screenshot(playwright.PageScreenshotOptions{
		Path:     playwright.String(path),
		FullPage: playwright.Bool(fullPage),
	})
	return err
}

// SnapshotEvents returns + clears the per-step event buffers.
func (b *Browser) SnapshotEvents() (cons []ConsoleEntry, net []NetworkEntry, errs []string) {
	cons = b.ConsoleMessages
	net = b.FailedRequests
	errs = b.PageErrors
	b.ConsoleMessages = nil
	b.FailedRequests = nil
	b.PageErrors = nil
	return
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}
